File: sdks/python/src/kafka_do/producer.py
"""
Kafka Producer implementation for kafka-do SDK.

Provides an async producer with kafka-python compatible API
using rpc-do for the underlying transport.
"""

from __future__ import annotations

import base64
from typing import TYPE_CHECKING, Any

from .types import (
    BatchResult,
    CompressionType,
    KafkaConnectionError,
    KafkaError,
    ProducerError,
    ProducerRecord,
    RecordMetadata,
    deserialize_record_metadata,
    serialize_record,
)

if TYPE_CHECKING:
    from .kafka import Kafka


class Producer:
    """
    Async Kafka producer for sending messages to Kafka topics.

    The producer connects to a Kafka service via rpc-do and provides
    methods for sending individual messages or batches.

    Example:
        kafka = Kafka(brokers=['localhost:9092'])
        producer = kafka.producer()
        await producer.connect()

        # Send a single message
        metadata = await producer.send('my-topic', value=b'Hello!')

        # Send with key
        metadata = await producer.send('my-topic', key=b'key1', value=b'message')

        # Send batch
        result = await producer.send_batch([
            {'topic': 'topic1', 'value': b'msg1'},
            {'topic': 'topic2', 'value': b'msg2'}
        ])

        await producer.disconnect()
    """

    __slots__ = (
        "_kafka",
        "_client",
        "_connected",
        "_acks",
        "_compression_type",
        "_batch_size",
        "_linger_ms",
        "_client_id",
    )

    def __init__(
        self,
        kafka: Kafka,
        *,
        acks: int | str = 1,
        compression_type: CompressionType = CompressionType.NONE,
        batch_size: int = 16384,
        linger_ms: int = 0,
        client_id: str | None = None,
    ) -> None:
        """
        Initialize the producer.

        Args:
            kafka: The parent Kafka instance
            acks: Number of acknowledgments required (0, 1, or 'all')
            compression_type: Compression type for messages
            batch_size: Maximum batch size in bytes
            linger_ms: Time to wait before sending batched messages
            client_id: Optional client identifier
        """
        self._kafka = kafka
        self._client: Any = None
        self._connected = False
        self._acks = acks
        self._compression_type = compression_type
        self._batch_size = batch_size
        self._linger_ms = linger_ms
        self._client_id = client_id

    @property
    def connected(self) -> bool:
        """Return whether the producer is connected."""
        return self._connected

    async def connect(self) -> None:
        """
        Connect to the Kafka service.

        Raises:
            KafkaConnectionError: If connection fails
        """
        if self._connected:
            return

        try:
            self._client = await self._kafka._get_client()
            self._connected = True
        except Exception as e:
            raise KafkaConnectionError(f"Failed to connect producer: {e}") from e

    async def disconnect(self) -> None:
        """
        Disconnect from the Kafka service.

        This will flush any pending messages before disconnecting.
        """
        if not self._connected:
            return

        try:
            await self.flush()
        finally:
            self._connected = False
            self._client = None

    async def send(
        self,
        topic: str,
        value: bytes,
        key: bytes | None = None,
        partition: int | None = None,
        timestamp: int | None = None,
        headers: list[tuple[str, bytes]] | None = None,
    ) -> RecordMetadata:
        """
        Send a single message to a Kafka topic.

        Args:
            topic: The topic to send to
            value: The message value
            key: Optional message key (used for partitioning)
            partition: Optional specific partition to send to
            timestamp: Optional message timestamp (milliseconds since epoch)
            headers: Optional message headers

        Returns:
            RecordMetadata with information about the sent message

        Raises:
            KafkaConnectionError: If not connected
            ProducerError: If send fails
        """
        if not self._connected:
            raise KafkaConnectionError("Producer not connected")

        record = ProducerRecord(
            topic=topic,
            value=value,
            key=key,
            partition=partition,
            timestamp=timestamp,
            headers=headers or [],
        )

        try:
            result = await self._client.kafka.producer.send(serialize_record(record))
            return deserialize_record_metadata(result)
        except Exception as e:
            raise ProducerError(f"Failed to send message: {e}") from e

    async def send_batch(
        self,
        records: list[dict[str, Any] | ProducerRecord],
    ) -> BatchResult:
        """
        Send a batch of messages.

        Args:
            records: List of records to send. Each record can be either:
                - A ProducerRecord instance
                - A dict with keys: topic, value, and optionally key, partition, timestamp, headers

        Returns:
            BatchResult with successful and failed records

        Raises:
            KafkaConnectionError: If not connected
        """
        if not self._connected:
            raise KafkaConnectionError("Producer not connected")

        # Convert dicts to ProducerRecords
        producer_records: list[ProducerRecord] = []
        for record in records:
            if isinstance(record, ProducerRecord):
                producer_records.append(record)
            else:
                producer_records.append(
                    ProducerRecord(
                        topic=record["topic"],
                        value=record["value"],
                        key=record.get("key"),
                        partition=record.get("partition"),
                        timestamp=record.get("timestamp"),
                        headers=record.get("headers", []),
                    )
                )

        # Serialize all records
        serialized = [serialize_record(r) for r in producer_records]

        try:
            results = await self._client.kafka.producer.sendBatch(serialized)
            batch_result = BatchResult()

            for i, result in enumerate(results):
                if result.get("error"):
                    error = KafkaError(
                        result["error"].get("message", "Unknown error"),
                        code=result["error"].get("code"),
                    )
                    batch_result.failed.append((producer_records[i], error))
                else:
                    batch_result.successful.append(deserialize_record_metadata(result))

            return batch_result
        except Exception as e:
            # If the entire batch fails, mark all as failed
            batch_result = BatchResult()
            error = ProducerError(f"Batch send failed: {e}")
            for record in producer_records:
                batch_result.failed.append((record, error))
            return batch_result

    async def flush(self, timeout: float | None = None) -> None:
        """
        Flush any pending messages.

        Args:
            timeout: Maximum time to wait for flush (None for default)
        """
        if not self._connected:
            return

        try:
            await self._client.kafka.producer.flush(timeout=timeout)
        except Exception:
            # Ignore flush errors on disconnect
            pass

    async def partitions_for(self, topic: str) -> list[int]:
        """
        Get the list of partitions for a topic.

        Args:
            topic: The topic name

        Returns:
            List of partition numbers

        Raises:
            KafkaConnectionError: If not connected
            KafkaError: If the operation fails
        """
        if not self._connected:
            raise KafkaConnectionError("Producer not connected")

        try:
            result = await self._client.kafka.producer.partitionsFor(topic)
            return list(result)
        except Exception as e:
            raise KafkaError(f"Failed to get partitions: {e}") from e

    async def __aenter__(self) -> Producer:
        """Async context manager entry."""
        await self.connect()
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: Any,
    ) -> None:
        """Async context manager exit."""
        await self.disconnect()

File: sdks/ruby/lib/kafka_do/producer.rb
# frozen_string_literal: true

require 'base64'
require 'json'

module KafkaDo
  # Kafka message producer
  #
  # Buffers messages and sends them in batches for efficiency.
  # Follows the ruby-kafka Producer API.
  #
  # @example Basic usage
  #   producer = kafka.producer
  #   producer.produce('Hello!', topic: 'my-topic')
  #   producer.produce('World!', topic: 'my-topic', key: 'greeting')
  #   producer.deliver_messages
  #
  # @example With block (auto-delivers)
  #   kafka.producer do |producer|
  #     producer.produce('Hello!', topic: 'my-topic')
  #   end
  #
  class Producer
    # @return [Kafka] the parent Kafka client
    attr_reader :kafka

    # @return [Integer] the number of buffered messages
    attr_reader :buffer_size

    # @return [Integer] the byte size of buffered messages
    attr_reader :buffer_bytesize

    # Create a new Producer
    #
    # @param kafka [Kafka] the parent Kafka client
    # @param compression_codec [Symbol] compression type
    # @param required_acks [Integer] number of acks required
    # @param max_buffer_size [Integer] maximum buffer size
    # @param max_buffer_bytesize [Integer] maximum buffer byte size
    def initialize(kafka, compression_codec: :none, required_acks: 1,
                   max_buffer_size: 1000, max_buffer_bytesize: 10_000_000, **_options)
      @kafka = kafka
      @compression_codec = compression_codec
      @required_acks = required_acks
      @max_buffer_size = max_buffer_size
      @max_buffer_bytesize = max_buffer_bytesize
      @buffer = []
      @buffer_size = 0
      @buffer_bytesize = 0
      @closed = false
      @mutex = Mutex.new
    end

    # Produce a message to a topic
    #
    # The message is buffered and will be sent when deliver_messages is called
    # or when the buffer is full.
    #
    # @param value [String, Hash] the message value (Hash will be JSON encoded)
    # @param topic [String] the target topic
    # @param key [String, nil] optional message key for partitioning
    # @param partition [Integer, nil] specific partition (auto-assigned if nil)
    # @param partition_key [String, nil] key used for partition calculation
    # @param headers [Hash, nil] optional message headers
    # @param create_time [Time, nil] optional message timestamp
    # @return [self] for method chaining
    #
    # @example Produce with value only
    #   producer.produce('Hello!', topic: 'greetings')
    #
    # @example Produce with key
    #   producer.produce('Hello!', topic: 'greetings', key: 'en')
    #
    # @example Produce with headers
    #   producer.produce('Hello!', topic: 'greetings',
    #     headers: { 'correlation-id' => 'abc-123' })
    #
    def produce(value, topic:, key: nil, partition: nil, partition_key: nil,
                headers: nil, create_time: nil)
      raise ClosedError, 'Producer is closed' if @closed

      # Encode value if it's a Hash
      encoded_value = value.is_a?(Hash) ? JSON.generate(value) : value.to_s
      encoded_key = key&.to_s

      message = {
        topic: topic,
        value: Base64.strict_encode64(encoded_value),
        partition: partition || -1
      }

      message[:key] = Base64.strict_encode64(encoded_key) if encoded_key
      message[:partition_key] = partition_key if partition_key
      message[:timestamp] = (create_time || Time.now).to_i * 1000

      if headers && !headers.empty?
        message[:headers] = headers.map do |k, v|
          { key: k.to_s, value: Base64.strict_encode64(v.to_s) }
        end
      end

      bytesize = encoded_value.bytesize + (encoded_key&.bytesize || 0)

      @mutex.synchronize do
        @buffer << message
        @buffer_size += 1
        @buffer_bytesize += bytesize

        # Auto-deliver if buffer is full
        if @buffer_size >= @max_buffer_size || @buffer_bytesize >= @max_buffer_bytesize
          deliver_messages_internal
        end
      end

      self
    end

    # Deliver all buffered messages
    #
    # Sends all buffered messages to Kafka and clears the buffer.
    #
    # @return [Array<RecordMetadata>] metadata for each delivered message
    # @raise [ProducerError] if delivery fails
    #
    # @example
    #   producer.produce('Hello!', topic: 'my-topic')
    #   producer.produce('World!', topic: 'my-topic')
    #   results = producer.deliver_messages
    #   results.each { |r| puts "Sent to partition #{r.partition}" }
    #
    def deliver_messages
      @mutex.synchronize { deliver_messages_internal }
    end

    # Clear the buffer without sending messages
    #
    # @return [void]
    def clear_buffer
      @mutex.synchronize do
        @buffer.clear
        @buffer_size = 0
        @buffer_bytesize = 0
      end
    end

    # Check if the buffer is empty
    #
    # @return [Boolean]
    def buffer_empty?
      @buffer_size.zero?
    end

    # Shutdown the producer, delivering any remaining messages
    #
    # @return [void]
    def shutdown
      deliver_messages unless buffer_empty?
      @closed = true
    end

    # Close the producer without delivering remaining messages
    #
    # @return [void]
    def close
      clear_buffer
      @closed = true
    end

    # Check if the producer is closed
    #
    # @return [Boolean]
    def closed?
      @closed
    end

    private

    def deliver_messages_internal
      return [] if @buffer.empty?

      messages = @buffer.dup
      @buffer.clear
      @buffer_size = 0
      @buffer_bytesize = 0

      # Group messages by topic for efficient batching
      by_topic = messages.group_by { |m| m[:topic] }

      results = []

      by_topic.each do |topic, topic_messages|
        batch_results = send_batch(topic, topic_messages)
        results.concat(batch_results)
      end

      results
    end

    def send_batch(topic, messages)
      client = kafka.rpc_client

      promise = client.call('kafka.producer.sendBatch', {
                              topic: topic,
                              messages: messages,
                              compression: @compression_codec.to_s,
                              acks: acks_value
                            })

      result = promise.await
      parse_batch_result(result, messages)
    rescue StandardError => e
      raise ProducerError.new("Failed to deliver messages: #{e.message}", cause: e)
    end

    def acks_value
      case @required_acks
      when :all, -1 then 'all'
      when 0 then '0'
      else '1'
      end
    end

    def parse_batch_result(result, messages)
      return [] unless result.is_a?(Array)

      result.map.with_index do |r, i|
        if r.is_a?(Hash) && r['error']
          raise ProducerError.new(r['error']['message'] || 'Send failed',
                                  code: r['error']['code'])
        end

        RecordMetadata.from_rpc(r || { 'topic' => messages[i][:topic] })
      end
    end
  end

  # Async producer for non-blocking message delivery
  #
  # @example
  #   producer = kafka.async_producer('orders')
  #   producer.produce('Hello!', topic: 'my-topic')
  #   producer.deliver
  #
  class AsyncProducer < Producer
    # Callback for delivery reports
    attr_accessor :on_delivery_callback

    def initialize(kafka, **options)
      super
      @pending_deliveries = []
      @on_delivery_callback = nil
    end

    # Set a callback for delivery reports
    #
    # @yield [RecordMetadata] called when a message is delivered
    def on_delivery(&block)
      @on_delivery_callback = block
    end

    # Trigger delivery of buffered messages (non-blocking)
    #
    # @return [void]
    def deliver
      Thread.new { deliver_messages }
    end
  end
end
